cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,153 @@
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        parsed_url = urlparse(url)
+        if parsed_url.scheme != "https":
+            raise ValueError("GIT dependencies must use HTTPS protocol")
+
+        return url
+
+    @pydantic.field_validator("revision")
+    def validate_git_revision(cls, revision: str) -> str:
+        """Validate that the git revision is a valid git ref."""
+        if not re.match(r"^[a-fA-F0-9]{40}$", revision):
+            raise ValueError("GIT dependencies must have a valid revision")
+
+        return revision
+
+
+class PathDependency(_GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this path dependency."""
+        return NotImplemented
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+ParseResult = list[BundlerDependency]
+
+
+def parse_lockfile(package_dir: RootedPath) -> ParseResult:
+    """Parse a Gemfile.lock file and return a list of dependencies."""
+    lockfile_path = package_dir.join_within_root(GEMFILE_LOCK)
+    gemfile_path = package_dir.join_within_root(GEMFILE)
+    if not lockfile_path.path.exists() or not gemfile_path.path.exists():
+        reason = "Gemfile and Gemfile.lock must be present in the package directory"
+        solution = (
+            "Run `bundle init` to generate the Gemfile.\n"
+            "Run `bundle lock` to generate the Gemfile.lock."
+        )
+        raise PackageRejected(reason=reason, solution=solution)
+
+    scripts_dir = Path(__file__).parent / "scripts"
+    lockfile_parser = scripts_dir / "lockfile_parser.rb"
+    try:
+        output = run_cmd(cmd=[str(lockfile_parser)], params={"cwd": package_dir.path})
+    except subprocess.CalledProcessError:
+        raise PackageManagerError(f"Failed to parse {lockfile_path}")
+
+    json_output = json.loads(output)
+
+    bundler_version: str = json_output["bundler_version"]
+    log.info("Package %s is bundled with version %s", package_dir.path.name, bundler_version)
+    dependencies: list[dict[str, str]] = json_output["dependencies"]
+
+    result: ParseResult = []
+    for dep in dependencies:
+        if dep["type"] == "rubygems":
+            result.append(GemDependency.model_validate(dep))
+        elif dep["type"] == "git":
+            result.append(GitDependency.model_validate(dep))
+        elif dep["type"] == "path":
+            _validate_path_dependency_subpath(package_dir, dep["path"])
+            result.append(PathDependency.model_validate(dep))
```

Comment:
Another observation (non-blocking): I think the validator could be tucked into `PathDependency`. This would require a factory to produce `PathDependency` subclasses with a bound `package_dir`. Then all three branches will be neatly uniform.